agent/ebs/watcher.go

```diff
@@ -0,0 +1,212 @@
+			log.Warnf("Failed to handle resource attachment %v", ebs.String())
+		}
+		if empty && len(w.agentState.GetAllPendingEBSAttachments()) == 1 {
+			w.scanTicker.Stop()
+			w.scanTicker = time.NewTicker(apiebs.ScanPeriod)
+		}
+	}
+	wg.Wait()
+	return err
+}
+
+// handleEBSAttachment will handle an EBS attachment via the following:
+// 1. Check whether we already have this attachment in state, if so, return
+// 2. Otherwise add the attachment to state, start its ack timer, and save the state
+func (w *EBSWatcher) handleEBSAttachment(ebs *apiebs.ResourceAttachment) error {
+	if ebs.AttachmentProperties[apiebs.ResourceTypeName] != apiebs.ElasticBlockStorage {
+		log.Warn("Resource type not Elastic Block Storage. Skip handling resource attachment.")
+		return nil
+	}
+	volumeID := ebs.AttachmentProperties[apiebs.VolumeIdName]
+	_, ok := w.agentState.GetEBSByVolumeId(volumeID)
+	if ok {
+		log.Infof("EBS Volume attachment already exists. Skip handling EBS attachment %v.", ebs.String())
+		return nil
+	}
+
+	if err := w.addEBSAttachmentToState(ebs); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (w *EBSWatcher) NotifyFound(foundVolumes []string) {
+	for _, volumeId := range foundVolumes {
+		w.notifyFoundEBS((volumeId))
+	}
+}
+
+// notifyFoundEBS will mark it as found within the agent state
+func (w *EBSWatcher) notifyFoundEBS(volumeId string) {
+	w.mailbox <- func() {
+		ebs, ok := w.agentState.GetEBSByVolumeId(volumeId)
+		if !ok {
+			log.Warnf("Unable to find EBS volume with volume ID: %v.", volumeId)
+			return
+		}
+
+		if ebs.HasExpired() {
+			log.Warnf("EBS status expired, no longer tracking EBS volume: %v.", ebs.String())
+			return
+		}
+
+		if ebs.IsSent() {
+			log.Warnf("State change event has already been emitted for EBS volume: %v.", ebs.String())
+			return
+		}
+
+		if ebs.IsAttached() {
+			log.Infof("EBS volume: %v, has already been found. Skip notifying EBS volume.", ebs.String())
+			return
+		}
+
+		ebs.StopAckTimer()
+		ebs.SetAttachedStatus()
+
+		log.Infof("Successfully found attached EBS volume: %v", ebs.String())
+		if len(w.agentState.GetAllPendingEBSAttachments()) == 0 {
+			log.Info("No more attachments to scan for. Stopping scan ticker.")
+			w.scanTicker.Stop()
+		}
+	}
+}
+
+// RemoveAttachment will stop tracking an EBS attachment
+func (w *EBSWatcher) RemoveAttachment(volumeID string) {
+	w.mailbox <- func() {
+		w.removeEBSAttachment(volumeID)
+	}
+}
+
+func (w *EBSWatcher) removeEBSAttachment(volumeID string) {
+	// TODO: Remove the EBS volume from the data client.
+	w.agentState.RemoveEBSAttachment(volumeID)
+	if len(w.agentState.GetAllPendingEBSAttachments()) == 0 {
+		log.Info("No more attachments to scan for. Stopping scan ticker.")
+		w.scanTicker.Stop()
+	}
+}
+
+// addEBSAttachmentToState adds an EBS attachment to state, and start its ack timer
+func (w *EBSWatcher) addEBSAttachmentToState(ebs *apiebs.ResourceAttachment) error {
+	volumeId := string(ebs.AttachmentProperties[apiebs.VolumeIdName])
```

Comment:
Is `ebs.AttachmentProperties[apiebs.VolumeIdName]` not of type `string`? Curious about the casting.

Reply:
Good call out. Fixed!